crates/ra_flycheck/src/lib.rs

```diff
@@ -178,17 +173,18 @@ impl FlycheckThread {
         }
     }
 
-    fn handle_message(&mut self, msg: CheckEvent, task_send: &Sender<CheckTask>) {
+    fn handle_message(&self, msg: CheckEvent, task_send: &Sender<CheckTask>) {
         match msg {
             CheckEvent::Begin => {
-                self.progress = Some(self.progress_src.begin(()));
+                task_send.send(CheckTask::Status(Status::Being)).unwrap();
+            }
+
+            CheckEvent::End => {
+                task_send.send(CheckTask::Status(Status::End)).unwrap();
             }
```

Comment:
This has changed the behavior of the original PR. `Progress` is a hande, overwriting it cancel currently active progress, e.g. when the user saved the file and flycheck started executing `cargo check` but then she the file again shortly we need to free the currently active progress resource.
@matklad that was the main reason to extract `Progress` handle, not to leak progress resources.